fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java

```diff
@@ -0,0 +1,343 @@
+
+import com.codahale.metrics.Timer;
+
+/**
+ * External binary from a url
+ *
+ * @author bbpennel
+ * @since 12/14/2017
+ */
+public class UrlBinary extends AbstractFedoraBinary {
+    private static final Logger LOGGER = getLogger(UrlBinary.class);
+
+    /**
+     * Construct UrlBinary
+     *
+     * @param node node
+     */
+    public UrlBinary(final Node node) {
+        super(node);
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see org.fcrepo.kernel.modeshape.FedoraBinaryImpl#getContent()
+     */
+    @Override
+    public InputStream getContent() {
+        // todo - this needs to be more complete so the proxy information will
+        // make it up to the higher levels. Ie, so one can pass back the response information
+        try {
+            return getResourceUri().toURL().openStream();
+        } catch (final IOException e) {
+            throw new RepositoryRuntimeException(e);
+        }
+    }
+
+    @Override
+    public long getContentSize() {
+        final long sizeValue = super.getContentSize();
+        if (sizeValue > -1L) {
+            return sizeValue;
+        }
+        try {
+            final String content = IOUtils.toString(getResourceUri().toURL().openStream());
+            setContentSize(content.length());
+            return content.length();
+        } catch (final IOException e) {
+            LOGGER.warn("Error getting getContentSize for '{}' : '{}'", getPath(), getResourceUri());
+            // Error getting remote size.
+        }
+        return -1L;
+    }
+
+    @Override
+    public void setExternalContent(final InputStream content, final String contentType,
+                           final Collection<URI> checksums, final String originalFileName,
+                           final String externalHandling, final String externalUrl)
+            throws InvalidChecksumException {
+
+        // set a few things on the description node, then set a few more in the other setContent() function
+        final Node binaryNode = getNode();
+        try {
+            if (externalHandling.equals(PROXY)) {
+                binaryNode.setProperty(PROXY_FOR, externalUrl);
+            } else if (externalHandling.equals(REDIRECT)) {
+                binaryNode.setProperty(REDIRECTS_TO, externalUrl);
+            } else {
+                throw new RepositoryException("Unknown external content handling type: " + externalHandling);
+            }
+        } catch (final RepositoryException e) {
+            throw new RepositoryRuntimeException(e);
+        }
+
+        setContent(content, contentType, checksums, originalFileName, null);
+    }
+    /*
+     * (non-Javadoc)
+     * @see org.fcrepo.kernel.modeshape.FedoraBinaryImpl#setContent(java.io.InputStream, java.lang.String,
+     * java.util.Collection, java.lang.String, org.fcrepo.kernel.api.services.policy.StoragePolicyDecisionPoint)
+     */
+    @Override
+    public void setContent(final InputStream content, final String contentType,
```

Comment:
In my opinion, setContent for URLBinary's should throw a `UnsupportedOperationException` with a message about using setExternalContent instead. The relevant parts of this method (mainly decorating properties) should be moved up to setExternalContent.  We dont' need to create a jcr:data anymore, so that can be dropped.

Reply:
Okay. I'm in agreement - I'll try this out to see what ripple effect it has... :)